python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,234 @@
+    tags = {"ray-node-status": status_tag}
+
+    if labels["ray.io/node-type"] == "head":
+        tags["ray-node-type"] = "head"
+        tags["ray-user-node-type"] = "head-group"
+    else:
+        tags["ray-node-type"] = "worker"
+        tags["ray-user-node-type"] = labels["ray.io/group"]
+
+    return tags
+
+
+class KuberayNodeProvider(NodeProvider):  # type: ignore
+    def __init__(
+        self,
+        provider_config: Dict[str, Any],
+        cluster_name: str,
+        _allow_multiple: bool = False,
+    ):
+        logger.info("Creating ClientNodeProvider.")
+        self.namespace = provider_config["namespace"]
+        self.cluster_name = cluster_name
+        self._lock = threading.RLock()
+
+        with open("/var/run/secrets/kubernetes.io/serviceaccount/token") as secret:
+            token = secret.read()
+
+        self.headers = {
+            "Authorization": "Bearer " + token,
+        }
+        self.verify = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+        # Disallow multiple node providers, unless explicitly allowed for testing.
+        global provider_exists
+        if not _allow_multiple:
+            assert (
+                not provider_exists
+            ), "Only one ClientNodeProvider allowed per process."
+        assert (
+            provider_config.get("disable_node_updaters", False) is True
+        ), "Must disable node updaters to use ClientNodeProvider."
+        provider_exists = True
+
+        super().__init__(provider_config, cluster_name)
+
+    def _get(self, path):
+        result = requests.get("https://kubernetes.default:443" + path, headers=self.headers, verify=self.verify)
+        assert result.status_code == 200
+        return result.json()
+
+    def _patch(self, path, payload):
+        result = requests.patch(
+            "https://kubernetes.default:443" + path,
+            json.dumps(payload),
+            headers={**self.headers, "Content-type": "application/json-patch+json"},
+            verify=self.verify
+        )
+        assert result.status_code == 200
+        return result.json()
+
+    def _get_worker_group(self, raycluster: Dict[str, Any], group_name: str) -> Tuple[int, Dict[str, str]]:
+        group_index = None
+        group_spec = None
+        worker_group_specs = raycluster["spec"]["workerGroupSpecs"]
+        for index, spec in enumerate(worker_group_specs):
+            if spec["groupName"] == group_name:
+                group_index = index
+                group_spec = spec
+        assert group_index is not None and group_spec is not None
+        return group_index, group_spec
+
+    def _wait_for_pods(self, group_name: str, replicas: int):
+        label_filters = to_label_selector({"ray.io/cluster": self.cluster_name, "ray.io/group": group_name})
+        while True:
+            pods = self._get("/api/v1/namespaces/default/pods?labelSelector=" + requests.utils.quote(label_filters))
```

Comment:
we have to take `namespace` into the consideration

Reply:
Yes totally, I will address that before merging this, thanks for spotting it :)

Reply:
this is fixed now